### PR TITLE
Handle esr and deep links

### PR DESCRIPTION
**Pull request by n13**
I kept deep links exactly as before

Only ESR links are also handled. 

I feel this is safest to make sure the post-install app store link still works, but we can only test that once we're live.

I tested 4 different scenarios
- Cold start with ESR
- Cold start with firebase link
- ESR when app is in the background but running
- Deep link when app in background but running

All 4 work - Cold start with firebase link has an issue on iPhone simulator, but works on device - tested in build 18 on Testflight. 

Test URLs

ESR
```
esr://gmNgYmBYlmzC9MoglIFhB9frlxK9jIwMEMAEpQVhAhyWppZGhhZGRowJGSUlBVb6-okFmboFiZW5qXklehmVBRmJeqmJRSUZ-iVFiXnFicklmfl59qWZKbaGiUlJySbmybqpKUbJuqaWFqm6FqlJJrqpSWYpaeYWiYZJSeZqJRXxQKXV1SUVtbUMAA
```

Firebase (not a valid invite code BTW so this won't actually be able to create an account)
```
https://hyphawallet.page.link/?link=https://hypha.earth/invite?code%3D12345abcde%26chain%3Dtelos%26dao%3D219987%26env%3Ddev&apn=earth.hypha.wallet.hypha_wallet&isi=1659926348&ibi=earth.hypha.wallet.hyphaWallet

```

### CODE REVIEW

**Comment by gguijarro-c-chwy** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>:exclamation: PR has more than 500 line changes </td>
    </tr>
  </tbody>
</table>



> Pull Request size seems relatively large. If Pull Request contains multiple changes, split each into separate PR will helps faster, easier review.
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against a87c80a34f45a8e0c05ad77807b752cff5b9ea68
</p>

File: lib/core/local/services/crypto_auth_service.dart [hunk 1]
```diff
@@ -1,5 +1,3 @@
-import 'dart:typed_data';
```
Comment (n13) [inline lib/core/local/services/crypto_auth_service.dart:1]: random cleanuo

File: lib/ui/blocs/deeplink/deeplink_bloc.dart [hunk 2]
```diff
@@ -37,10 +79,7 @@ class DeeplinkBloc extends Bloc<DeeplinkEvent, DeeplinkState> {
   Future<void> _incomingFirebaseDeepLink(_IncomingFirebaseDeepLink event, Emitter<DeeplinkState> emit) async {
     /// fetch data from link.
     final queryParams = event.link.queryParameters;
-    if (queryParams.isNotEmpty &&
-        queryParams.containsKey('code') &&
-        queryParams.containsKey('chain') &&
-        queryParams.containsKey('dao')) {
+    if (queryParams.isNotEmpty && queryParams.containsKey('code') && queryParams.containsKey('chain')) {
```
Comment (n13) [inline lib/ui/blocs/deeplink/deeplink_bloc.dart:82]: FYI "dao" is actually not mandatory. This could be in another PR. Since it has nothing to do with this one.

File: lib/ui/blocs/deeplink/deeplink_state.dart [hunk 3]
```diff
@@ -11,7 +11,7 @@ class DeeplinkState with _$DeeplinkState {
 class InviteLinkData {
```
Comment (n13) [inline lib/ui/blocs/deeplink/deeplink_state.dart:11]: This is the only change for non mandatory dao

File: macos/Flutter/GeneratedPluginRegistrant.swift [hunk 4]
```diff
@@ -5,6 +5,7 @@
 import FlutterMacOS
```
Comment (n13) [inline macos/Flutter/GeneratedPluginRegistrant.swift:5]: generated changes by flutter

File: lib/core/crypto/dart_esr/src/signing_request_manager.dart [hunk 5]
```diff
@@ -178,6 +178,12 @@ class SigningRequestManager {
     return SigningRequestManager.fromData(buf.asUint8List(), options: options);
   }
 
+  static bool isValidESRScheme(String uri) {
```
Comment (gguijarro-c-chwy) [inline lib/core/crypto/dart_esr/src/signing_request_manager.dart:181]: 👍

File: lib/ui/blocs/deeplink/deeplink_bloc.dart [hunk 6]
```diff
@@ -1,34 +1,76 @@
...
 
 part 'deeplink_bloc.freezed.dart';
 part 'deeplink_event.dart';
 part 'deeplink_state.dart';
 part 'page_command.dart';
 
 class DeeplinkBloc extends Bloc<DeeplinkEvent, DeeplinkState> {
-  DeeplinkBloc() : super(const DeeplinkState()) {
+  final ParseQRCodeUseCase _parseQRCodeUseCase;
+  final _appLinks = AppLinks();
```
Comment (gguijarro-c-chwy) [inline lib/ui/blocs/deeplink/deeplink_bloc.dart:20]: AppLinks() what is this?

File: lib/ui/blocs/deeplink/deeplink_bloc.dart [hunk 7]
```diff
@@ -1,34 +1,76 @@
...
+  DeeplinkBloc(this._parseQRCodeUseCase) : super(const DeeplinkState()) {
     initDynamicLinks();
 
     on<_IncomingFirebaseDeepLink>(_incomingFirebaseDeepLink);
+    on<_IncomingESRLink>(_incomingESRLink);
     on<_ClearPageCommand>((_, emit) => emit(state.copyWith(command: null)));
   }
 
+  void _debugShowMessage(String message) {
+    Get.showSnackbar(GetSnackBar(message: message, duration: const Duration(seconds: 5)));
```
Comment (gguijarro-c-chwy) [inline lib/ui/blocs/deeplink/deeplink_bloc.dart:33]: not in the bloc. Move to UI pls

File: lib/ui/blocs/deeplink/deeplink_bloc.dart [hunk 8]
```diff
@@ -54,4 +93,24 @@ class DeeplinkBloc extends Bloc<DeeplinkEvent, DeeplinkState> {
...
+  Future<void> _incomingESRLink(_IncomingESRLink event, Emitter<DeeplinkState> emit) async {
+    final result = await _parseQRCodeUseCase.run(ParseESRLinkInput(esrLink: event.link));
+    if (result.isValue) {
+      emit(
+        state.copyWith(
+          command: PageCommand.navigateToSignTransaction(result.asValue!.value),
+        ),
+      );
+    } else {
+      print('error: ${result.asError!.error}');
```
Comment (gguijarro-c-chwy) [inline lib/ui/blocs/deeplink/deeplink_bloc.dart:106]: long helper. If you copied this error printing from somewhere pls fix over there too.

File: pubspec.lock [hunk 9]
```diff
@@ -33,6 +33,14 @@ packages:
       url: "https://pub.dev"
     source: hosted
     version: "5.2.0"
+  app_links:
```
Comment (gguijarro-c-chwy) [inline pubspec.lock:36]: ahhi see, its a lib. ok

File: windows/flutter/generated_plugin_registrant.cc [hunk 10]
```diff
@@ -6,12 +6,15 @@
```
[the remark this answers is not in the document]
Comment (gguijarro-c-chwy) [reply windows/flutter/generated_plugin_registrant.cc:6]: Yes 100%

File: lib/ui/blocs/deeplink/deeplink_bloc.dart [hunk 11]
```diff
@@ -1,34 +1,76 @@
...
+      if (!initialDeepLink.link.isEsr()) {
+        add(DeeplinkEvent.incomingFirebaseDeepLink(initialDeepLink.link));
+      }
+    }
 
-    if (initialLink != null) {
-      LogHelper.d('initial link: ${initialLink.link}');
-      add(DeeplinkEvent.incomingFirebaseDeepLink(initialLink.link));
+    // Handle initial ESR links
+    final initialUriString = await _appLinks.getInitialAppLinkString();
```
Comment (gguijarro-c-chwy) [inline lib/ui/blocs/deeplink/deeplink_bloc.dart:48]: now that i know what app links is., 

Can we move the use of this AppLinks package into a service. Similar to Push notification service. 

If we need this PR ASAP, we can do on a diff PR.

File: lib/ui/blocs/deeplink/deeplink_bloc.dart (hunk 7 above)
Comment (n13) [reply lib/ui/blocs/deeplink/deeplink_bloc.dart:33]: Thats a debug leftover. Deleting.

File: lib/ui/blocs/deeplink/deeplink_bloc.dart (hunk 8 above)
Comment (n13) [reply lib/ui/blocs/deeplink/deeplink_bloc.dart]: Done

File: lib/ui/blocs/deeplink/deeplink_bloc.dart (hunk 11 above)
Comment (n13) [reply lib/ui/blocs/deeplink/deeplink_bloc.dart:44]: Not sure how to do that, it's a very small piece of code...

**Comment by n13** [comment]
Fixed some things

File: windows/flutter/generated_plugin_registrant.cc (hunk 10 above)
[the remark this answers is not in the document]
Comment (n13) [reply windows/flutter/generated_plugin_registrant.cc:6]: will do another pr
